app/src/main/java/org/olpc/xorobothash/MainActivity.java

```diff
@@ -0,0 +1,157 @@
+import android.view.View;
+import android.widget.ImageView;
+import android.util.Log;
+import android.widget.TextView;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.net.HttpURLConnection;
+
+public class MainActivity extends ActionBarActivity {
+
+    private ImageView robotImageView;
+
+    //For debugging
+    private static final String TAG = "XORobotHash";
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+        robotImageView = (ImageView) findViewById(R.id.robot_hash_image);
+        getRobot();
+    }
+
+    private void getRobot() {
+        boolean isConnected = checkNetworkConnectivity();
+        boolean done = updateView(isConnected);
+        if (!done) {
+           if(isConnected) {
+               // Will call updateView again when done
+               new DownloadRobotImage().execute(uniqueRoboHashURL(), robotImagePath());
+           }
+           // Nothing else to do, updateView already showed the right error message
+        }
+    }
+
+    private boolean updateView(boolean isConnected) {
+        Log.i(TAG, "In updateView()");
+        boolean robotVisible = false;
+        File file = new File(robotImagePath());
+        // Check if we have ever successfully downloaded the XO's robot image
+        if (file.exists()) {
+            Log.i(TAG, "Robohash image file exists.");
+            robotVisible = showRobot(robotImagePath());
+        } else if (!isConnected) {
+            Log.i(TAG, "Robohash image file does not exist, network is not connected.");
+            showDisconnected();
+        }
+        return robotVisible;
+    }
+
+    private boolean showRobot(String imagePath) {
+        Log.i(TAG, "In showRobot()");
+        boolean robotDrawn;
+        Bitmap bitmap = BitmapFactory.decodeFile(imagePath);
+        robotImageView.setImageBitmap(bitmap);
+        // Check if the robot image is visible
+        robotDrawn = robotImageView.getDrawable() == null ? false : true;
+        return robotDrawn;
+    }
+
+    private void showDisconnected() {
+        //Hide the robot text
+        TextView robotText = (TextView) findViewById(R.id.robot_text_message);
+        robotText.setVisibility(View.GONE);
+        //Hide the robot image
+        ImageView robotImage = (ImageView) findViewById(R.id.robot_hash_image);
+        robotImage.setVisibility(View.GONE);
+        //Show the disconnected error message
+        TextView errorMessage = (TextView) findViewById(R.id.error_message);
+        errorMessage.setText(getString(R.string.no_network_connection));
+    }
+
+    private String robotImagePath() {
+        return this.getFilesDir() + File.pathSeparator + "this-xos-robot.png";
+    }
+
+    private String uniqueRoboHashURL() {
```

Comment:
@anorth Do you suggest this because the method is named with `URL` or for another reason?

If I do return a `URL` here, all I can see happening is line 52 changing from:

`new DownloadRobotImage().execute(uniqueRoboHashURL(), robotImagePath());`
to
`new DownloadRobotImage().execute(uniqueRoboHashURL().toString(), robotImagePath());`